bin/lint-migrations-file/src/lint_migrations_file.clj

```diff
@@ -59,63 +83,76 @@
   "Ensures that no \"boolean\" type columns are added in changesets with id later than v49.00-032. From that point on,
   \"${boolean.type}\" should be used instead, so that we can consistently use `BIT(1)` for Boolean columns on MySQL."
   [change-log]
-  (assert-no-types-in-change-log #{"boolean"} change-log #(pos? (compare % "v49.00-032"))))
+  (require-no-types-in-change-log! #{"boolean"} change-log #(pos? (compare % "v49.00-032"))))
 
-(defn no-datetime-type?
+(defn require-no-datetime-type!
   "Ensures that no \"datetime\" or \"timestamp without time zone\".
   From that point on, \"${timestamp_type}\" should be used instead, so that all of our time related columsn are tz-aware."
   [change-log]
-  (assert-no-types-in-change-log
+  (require-no-types-in-change-log!
    #{"datetime" "timestamp" "timestamp without time zone"}
    change-log
    #(pos? (compare % "v49.00-000"))))
 
 (s/def ::changeSet
   (s/spec :change-set.strict/change-set))
 
-(s/def ::databaseChangeLog
-  (s/and distinct-change-set-ids?
-         change-set-ids-in-order?
-         no-bare-blob-or-text-types?
-         no-bare-boolean-types?
-         no-datetime-type?
-         (s/+ (s/alt :property              (s/keys :req-un [::property])
-                     :objectQuotingStrategy (s/keys :req-un [::objectQuotingStrategy])
-                     :changeSet             (s/keys :req-un [::changeSet])))))
-
-(defn- validate-migrations [migrations]
-  (when (= (s/conform ::migrations migrations) ::s/invalid)
-    (let [data (s/explain-data ::migrations migrations)]
-      (throw (ex-info (str "Validation failed:\n" (with-out-str (pprint/pprint (mapv #(dissoc % :val)
-                                                                                     (::s/problems data)))))
-                      (or (dissoc data ::s/value) {})))))
+(defn- only-key [m]
+  (let [keys (keys m)]
+    (when-not (= 1 (count keys))
+      (throw (validation-error "Expected exactly one key." {:keys keys})))
+    (first keys)))
+
+(defn- validate-database-change-log! [change-log]
+  (require-distinct-change-set-ids! change-log)
+  (require-change-set-ids-in-order! change-log)
+  (require-no-bare-blob-or-text-types! change-log)
+  (require-no-bare-boolean-types! change-log)
+  (require-no-datetime-type! change-log)
+  (let [{:keys [changeSet]
+         :as all} (group-by only-key change-log)]
+    (when-not (set/subset? (set (keys all)) #{:property :objectQuotingStrategy :changeSet})
+      (throw (validation-error "Expected exactly one of :property, :objectQuotingStrategy, or :changeSet."
+                               {:keys (keys all)})))
+    (doseq [{:keys [changeSet]} changeSet
+            :when (not (s/valid? ::changeSet changeSet))]
+      (throw (validation-error "Invalid change set." (s/explain-data ::changeSet changeSet))))))
+
+(defn- validate-migrations! [migrations]
+  (require-database-change-log! migrations)
+  (validate-database-change-log! (:databaseChangeLog migrations))
```

Comment:
Why use `!` in the function names? I'm used to that indicating that the function mutates state, is that the case here? (I don't see it)

Reply:
I think I've seen `!` as an indicator that the function's purpose is side effects in general (not just changing state), but that doesn't seem to be the case here - I'll change this back.